dataverse/webapi/C#-NETx/BulkOperations/UpsertMultiple/Program.cs

```diff
@@ -0,0 +1,349 @@
+                    // 2. Create x/4 records with a unique alternate key value
+                    // 3. Update x/4 records by providing Primary Key (PK) which is entity.Id
+                    // 4. Update x/4 records by providing Alternate Key (AK) which is entity.KeyAttribute
+
+                    // Create a List of entity instances to be updated during UpsertMultiple call.
+
+                    // To handle cases where number of records is an odd number
+                    var numRecCreate = Math.Ceiling((double)(numberOfRecords / 2)); 
+                    
+                    Console.WriteLine($"\nPreparing {numRecCreate} records to create..");
+
+                    List<JObject> entityList = new();
+
+                    // Populate the list with the number of records to test.
+                    for (int i = 0; i < numRecCreate; i++)
+                    {
+                        entityList.Add(new JObject() {
+                            {"sample_name", $"sample record {i+1:0000000}"},
+                            // Each record MUST have the @odata.type specified.
+                            {"@odata.type",$"Microsoft.Dynamics.CRM.{tableLogicalName}" }
+                        });
+                    }
+
+                    if(Settings.CreateAlternateKey)
+                    {
+                        for(int i = 0;i < numRecCreate; i++)
+                        {
+                            var entity = entityList[i];
+
+                            if (Settings.UseElastic)
+                            {
+                                entity.Add("@odata.id", 
+                                    $"{tableLogicalName}s({tablePrimaryKeyName}={Guid.NewGuid()}," +
+                                    $"{elasticTablePartitionId}='sample pre-upsert record key {i + 1:0000000}')");
+                            }
+                            else
+                            {
+                                // Add the alternate key in the payload in @odata.id tag
+                                entity.Add(new JProperty("@odata.id", 
+                                    $"{tableLogicalName}s({tableAlternateKey}='sample pre-upsert record key {i + 1:0000000}')"));
+                            }
+                        }
+                    }
+
+                    // Use UpsertMultiple
+                    UpsertMultipleRequest upsertMultipleRequest = new(
+                        entitySetName: tableSetName,
+                        targets: entityList);
+
+                    // Add a tag optional parameter to set a shared variable to be available to a plug-in.
+                    // See https://learn.microsoft.com/power-apps/developer/data-platform/optional-parameters?tabs=webapi#add-a-shared-variable-to-the-plugin-execution-context
+
+                    // TODO?: We could add a string 'tag' parameter to the UpsertMultipleRequest constructor and append it to the URL when it is provided.
```

Comment:
This 'tag' parameter is 'nice to have', but not essential. CreateMultiple and UpsertMultiple don't have it.

But it would be a cleaner way to set this IMO.
OK if you don't want to.